src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -755,7 +753,7 @@ public boolean hasntStartedYet() {
     }
 
     @Override public boolean isLogUpdated() {
-        return isBuilding(); // there is no equivalent to a post-production state for flows
+        return listener != null || isBuilding(); // there is no equivalent to a post-production state for flows
```

Comment:
IIUC, this is the main fix. Does `listener` need to be made `volatile` now to make sure that tests calling `assertLogContains` actually see the latest value?

Reply:
Maybe. I doubt any of our tests could observe anything close to the time resolution required for `volatile` to matter. Usually it is added to fields to ensure that observers do not see half-constructed objects, which is not relevant here since we are merely checking for nullness.